packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -33,70 +40,141 @@ function areDefinedAndEqual<T>(
   return (value1 || value2) !== undefined && value1 === value2;
 }
 
-function flattenResultTree(result: ResultWithFolding): ResultWithFolding[] {
-  const results = [result];
-  if (result.parentResult) {
-    results.unshift(result.parentResult);
-  }
-  if (result.childResults?.length) {
-    results.push(
-      ...result.childResults.flatMap((childResult) =>
-        flattenResultTree(childResult)
-      )
-    );
+function getChildResultsRecursively(
+  result: ResultWithFolding
+): ResultWithFolding[] {
+  if (!result.childResults) {
+    return [];
   }
-  return results;
+  return result.childResults.flatMap((childResult) => [
+    childResult,
+    ...getChildResultsRecursively(childResult),
+  ]);
 }
 
-function foldResult(
-  source: ResultWithFolding,
+function resolveChildrenFromFields(
+  parent: ResultWithFolding,
   results: ResultWithFolding[],
   fields: FoldingFields
-): FoldedResult {
-  const sourceChildValue = getChildField(source, fields);
-  return {
-    ...source,
-    children: sourceChildValue
-      ? results
-          .filter((result) => {
-            const isSameResultAsSource =
-              getChildField(result, fields) === getChildField(source, fields);
-            const isChildOfSource =
-              getParentField(result, fields) === sourceChildValue;
-            return isChildOfSource && !isSameResultAsSource;
-          })
-          .map((result) => foldResult(result, results, fields))
-      : [],
-  };
+): FoldedResult[] {
+  const sourceChildValue = getChildField(parent, fields);
+  return sourceChildValue
+    ? results
+        .filter((result) => {
+          const isSameResultAsSource =
+            getChildField(result, fields) === getChildField(parent, fields);
+          const isChildOfSource =
+            getParentField(result, fields) === sourceChildValue;
+          return isChildOfSource && !isSameResultAsSource;
+        })
+        .map((result) => ({
+          ...result,
+          children: resolveChildrenFromFields(result, results, fields),
+        }))
+    : [];
 }
 
-function createCollection(
-  relevantResult: ResultWithFolding,
+function resolveRootFromFields(
+  results: ResultWithFolding[],
   fields: FoldingFields
-): FoldedResult {
-  const flattenedResults = removeDuplicates(
-    flattenResultTree(relevantResult),
+) {
+  return results.find((result) => {
+    const hasNoParent = getParentField(result, fields) === undefined;
+    const isParentOfItself = areDefinedAndEqual(
+      getParentField(result, fields),
+      getChildField(result, fields)
+    );
+    return hasNoParent || isParentOfItself;
+  });
+}
+
+function getExpensiveFoldedResults(
```

Comment:
I wasn't sure what terminology to use. By "expensive" I mean results that count towards `filterFieldRange` and by "free" I mean results that don't count towards it.

To know whether there are more results available, we request one more folded result. However, occasionally, the search API may fill that slot with a parent result (which was already included in `parentResult` for free) even though no other results may be available in the collection. This would result in `moreResultsAvailable` being `true` but dispatching `loadCollection` making no difference.